### PR TITLE
[BOLT] Expose pseudo probe function checksum and GUID

### PR DESCRIPTION
**Pull request by aaupov**
Add a BinaryFunction field for pseudo probe function GUID.
Populate it during pseudo probe section parsing, and emit it in YAML
profile (both regular and BAT), along with function checksum.

To be used for stale function matching.

Test Plan: update pseudoprobe-decoding-inline.test

### CODE REVIEW

File: bolt/include/bolt/Core/BinaryFunction.h [hunk 1]
```diff
@@ -2256,6 +2260,11 @@ class BinaryFunction {
   /// Returns the last computed hash value of the function.
   size_t getHash() const { return Hash; }
 
+  /// Returns the function hash from pseudo-probe description of the function.
+  size_t getPseudoProbeDescHash() const { return PseudoProbeDescHash; }
```
Comment (wlei-llvm) [inline bolt/include/bolt/Core/BinaryFunction.h:2264]: Is this intentional to use `size_t`? looks all other places use  `uint64_t` for pseudo-probe hash.

Comment (aaupov) [reply bolt/include/bolt/Core/BinaryFunction.h:2264]: Good catch.

**Review by wlei-llvm — APPROVED** [review]
LGTM.

File: bolt/include/bolt/Core/BinaryContext.h [hunk 2]
```diff
@@ -377,6 +381,17 @@ class BinaryContext {
     RtLibrary = std::move(Lib);
   }
 
+  const MCPseudoProbeDecoder *getPseudoProbeDecoder() const {
+    return PseudoProbeDecoder.get();
+  }
+
+  MCPseudoProbeDecoder &
+  setPseudoProbeDecoder(std::unique_ptr<MCPseudoProbeDecoder> Decoder) {
```
Comment (WenleiHe) [inline bolt/include/bolt/Core/BinaryContext.h:389]: nit: it's a bit unconventional to have setter return a reference. Nearby code all have a setter with void return and a dedicated setter. Suggest to be consistent.

File: bolt/include/bolt/Core/BinaryFunction.h [hunk 3]
```diff
@@ -2256,6 +2260,11 @@ class BinaryFunction {
   /// Returns the last computed hash value of the function.
   size_t getHash() const { return Hash; }
 
+  /// Returns the function GUID from pseudo-probe description of the function.
+  uint64_t getPseudoProbeGUID() const { return PseudoProbeGUID; }
+
+  void setPseudoProbeGUID(uint64_t GUID) { PseudoProbeGUID = GUID; }
```
Comment (WenleiHe) [inline bolt/include/bolt/Core/BinaryFunction.h:2266]: This is function GUID, pseudo-probe doesn't really have a GUID. It's computed from function name: `Function::getGUID(FName)` Suggest rename to `s/PseudoProbe/Function/`for getter/setter/field.

**Comment by WenleiHe** [comment]
In theory we just need a function checksum, plus a set of stable block Ids for each function. I think we don't have to prefix everything with "pseudo-probe" -- probe is one way, and the current way of getting checksum and stable block id for llvm compiled binary when csspgo is used, but at bolt level, it doesn't have to be tied to probe and it can be anything that the compiler and binary provides.

**Comment by aaupov** [comment]
@WenleiHe: addressed your comments. Please let me know if the current state is good to go.

File: bolt/include/bolt/Profile/ProfileYAMLMapping.h [hunk 4]
```diff
@@ -151,6 +151,8 @@ struct BinaryFunctionProfile {
   llvm::yaml::Hex64 Hash{0};
   uint64_t ExecCount{0};
   std::vector<BinaryBasicBlockProfile> Blocks;
+  llvm::yaml::Hex64 GUID{0};
+  llvm::yaml::Hex64 PseudoProbeDescHash{0};
```
Comment (WenleiHe) [inline bolt/include/bolt/Profile/ProfileYAMLMapping.h:155]: This isn't really a pseudo-probe hash, just like GUID isn't really a pseudo-probe GUID. Can we call it CFG checksum/hash?

**Comment by WenleiHe** [comment]
> @WenleiHe: addressed your comments. Please let me know if the current state is good to go.

@aaupov one more comment related to what I mentioned above.

**Pull request merged**